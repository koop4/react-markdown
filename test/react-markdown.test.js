'use strict';

var React = require('react/addons'),
    jsdom = require('mocha-jsdom'),
    expect = require('chai').expect,
    ReactMarkdown = require('../');

var TestUtils = React.addons.TestUtils;

describe('ReactMarkdown', function() {
    jsdom();

    var testDate = (new Date()).toString();
    var testMarkdown = [
        '# Demo\n\n',
        'I was *rendered* using __React__ at ' + testDate,
        ' and it was so much FUN!', '\n\n',
        'Lets do it <i>again</i>!\nYeah?'
    ].join('');

    it('should not set a class on container if no className is passed as prop', function() {
        var rendered = TestUtils.renderIntoDocument(
            React.createElement(ReactMarkdown, { source: testMarkdown })
        );

        expect(rendered.getDOMNode().getAttribute('class')).to.equal(null);
    });

    it('should set a class on container if className is passed as prop', function() {
        var rendered = TestUtils.renderIntoDocument(
            React.createElement(ReactMarkdown, { source: testMarkdown, className: 'foo bar' })
        );

        expect(rendered.getDOMNode().getAttribute('class')).to.equal('foo bar');
    });

    it('should have rendered a div with the right children', function() {
        var rendered = TestUtils.renderIntoDocument(
            React.createElement(ReactMarkdown, { source: testMarkdown })
        );

        expect(rendered.getDOMNode().tagName).to.equal('DIV');
        expect(rendered.getDOMNode().innerHTML).to.contain(testDate);

        var h1 = TestUtils.findRenderedDOMComponentWithTag(rendered, 'h1');
        expect(h1.getDOMNode().innerHTML).to.equal('Demo');

        var em = TestUtils.findRenderedDOMComponentWithTag(rendered, 'em');
        expect(em.getDOMNode().innerHTML).to.equal('rendered');

        var strong = TestUtils.findRenderedDOMComponentWithTag(rendered, 'strong');
        expect(strong.getDOMNode().innerHTML).to.equal('React');

        var ps = TestUtils.scryRenderedDOMComponentsWithTag(rendered, 'p');
        expect(ps).to.have.length(2);
    });

    it('can specify different container tag name', function() {
        var rendered = TestUtils.renderIntoDocument(
            React.createElement(ReactMarkdown, { source: testMarkdown, containerTagName: 'section' })
        );

        expect(rendered.getDOMNode().tagName).to.equal('SECTION');
    });

    it('can be told to use <br> for soft-breaks', function() {
        var rendered = TestUtils.renderIntoDocument(
            React.createElement(ReactMarkdown, { source: testMarkdown, softBreak: 'br' })
        );

        var strong = TestUtils.findRenderedDOMComponentWithTag(rendered, 'br');
        expect(strong).to.be.ok;
    });

    it('can be told to output sourcemaps as data attributes', function() {
        var rendered = TestUtils.renderIntoDocument(
            React.createElement(ReactMarkdown, { source: testMarkdown, sourcePos: true })
        );

        var h1 = TestUtils.findRenderedDOMComponentWithTag(rendered, 'h1');
        expect(h1.getDOMNode().getAttribute('data-sourcepos')).to.equal('1:1-1:6');
    });

    it('can be told to escape html', function() {
        var rendered = TestUtils.renderIntoDocument(
            React.createElement(ReactMarkdown, { source: testMarkdown, escapeHtml: true })
        );

        var ps = TestUtils.scryRenderedDOMComponentsWithTag(rendered, 'p');
        expect(ps[1].getDOMNode().innerHTML).to.contain('&lt;i&gt;');
    });

    it('can be told to skip HTML', function() {
        var rendered = TestUtils.renderIntoDocument(
            React.createElement(ReactMarkdown, { source: testMarkdown, skipHtml: true })
        );

        var ps = TestUtils.scryRenderedDOMComponentsWithTag(rendered, 'p');
        expect(ps[1].getDOMNode().innerHTML).to.not.contain('<i>');
        expect(ps[1].getDOMNode().innerHTML).to.not.contain('&lt;i&gt;');
    });

    it('does not blow up if no source is given', function() {
        var rendered = TestUtils.renderIntoDocument(
            React.createElement(ReactMarkdown, {})
        );

        expect(rendered.getDOMNode().innerHTML).to.equal('');
    });
});
