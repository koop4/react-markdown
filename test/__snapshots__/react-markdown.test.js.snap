// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can render the most basic of documents (single paragraph) 1`] = `
<div>
  <p>
    Test
  </p>
</div>
`;

exports[`should be able to override renderers 1`] = `
<div>
  <span
    className="heading level-1"
  >
    Header
  </span>
  <p>
    Paragraph
  </p>
  <span
    className="heading level-2"
  >
    New header
  </span>
  <ol>
    <li>
      List item
    </li>
    <li>
      List item 2
    </li>
  </ol>
  <p>
    Foo
  </p>
</div>
`;

exports[`should be able to render inline code 1`] = `
<div>
  <p>
    Just call 
    <code>
      renderToStaticMarkup()
    </code>
    , already
  </p>
</div>
`;

exports[`should be able to render inline html 1`] = `
<div>
  <p>
    I am having 
    <span
      dangerouslySetInnerHTML={
        Object {
          "__html": "<strong>",
        }
      }
    />
    so
    <span
      dangerouslySetInnerHTML={
        Object {
          "__html": "</strong>",
        }
      }
    />
     much fun
  </p>
</div>
`;

exports[`should escape html blocks by default 1`] = `
<div>
  <p>
    This is a regular paragraph.
  </p>
  <span>
    &lt;table&gt;
    &lt;tr&gt;
        &lt;td&gt;Foo&lt;/td&gt;
    &lt;/tr&gt;
&lt;/table&gt;
  </span>
  <p>
    This is another regular paragraph.
  </p>
</div>
`;

exports[`should handle blockquotes 1`] = `
<div>
  <blockquote>
    <p>
      Moo
Tools
FTW
    </p>
  </blockquote>
</div>
`;

exports[`should handle bold/strong text 1`] = `
<div>
  <p>
    React is 
    <strong>
      totally
    </strong>
     
    <strong>
      awesome
    </strong>
  </p>
</div>
`;

exports[`should handle code tags with language specification 1`] = `
<div>
  <pre>
    <code
      className="language-js"
    >
      var foo = require('bar');
foo();
    </code>
  </pre>
</div>
`;

exports[`should handle code tags without any language specification 1`] = `
<div>
  <pre>
    <code>
      var foo = require('bar');
foo();
    </code>
  </pre>
</div>
`;

exports[`should handle emphasis 1`] = `
<div>
  <p>
    React is 
    <em>
      totally
    </em>
     
    <em>
      awesome
    </em>
  </p>
</div>
`;

exports[`should handle horizontal rules 1`] = `
<div>
  <p>
    Foo
  </p>
  <hr />
  <p>
    Bar
  </p>
</div>
`;

exports[`should handle html blocks 1`] = `
<div>
  <p>
    This is a regular paragraph.
  </p>
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "<table>
    <tr>
        <td>Foo</td>
    </tr>
</table>",
      }
    }
  />
  <p>
    This is another regular paragraph.
  </p>
</div>
`;

exports[`should handle images with custom uri transformer 1`] = `
<div>
  <p>
    This is 
    <img
      alt="an image"
      src="/ninja.jpg"
    />
    .
  </p>
</div>
`;

exports[`should handle images with special characters in alternative text 1`] = `
<div>
  <p>
    This is 
    <img
      alt="a ninja's image"
      src="/ninja.png"
    />
    .
  </p>
</div>
`;

exports[`should handle images with title attribute 1`] = `
<div>
  <p>
    This is 
    <img
      alt="an image"
      src="/ninja.png"
      title="foo bar"
    />
    .
  </p>
</div>
`;

exports[`should handle images without title attribute 1`] = `
<div>
  <p>
    This is 
    <img
      alt="an image"
      src="/ninja.png"
    />
    .
  </p>
</div>
`;

exports[`should handle inline html with escapeHtml option enabled 1`] = `
<div>
  <p>
    I am having 
    <span>
      &lt;strong&gt;
    </span>
    so
    <span>
      &lt;/strong&gt;
    </span>
     much fun
  </p>
</div>
`;

exports[`should handle links with custom uri transformer 1`] = `
<div>
  <p>
    This is 
    <a
      href="//espen.codes/"
    >
      a link
    </a>
     to Espen.Codes.
  </p>
</div>
`;

exports[`should handle links with title attribute 1`] = `
<div>
  <p>
    This is 
    <a
      href="https://espen.codes/"
      title="some title"
    >
      a link
    </a>
     to Espen.Codes.
  </p>
</div>
`;

exports[`should handle links without title attribute 1`] = `
<div>
  <p>
    This is 
    <a
      href="https://espen.codes/"
    >
      a link
    </a>
     to Espen.Codes.
  </p>
</div>
`;

exports[`should handle loose, unordered lists 1`] = `
<div>
  <ul>
    <li>
      <p>
        foo
      </p>
    </li>
    <li>
      bar
    </li>
  </ul>
</div>
`;

exports[`should handle multiline paragraphs properly (softbreak, paragraphs) 1`] = `
<div>
  <p>
    React is awesome
And so is markdown
    <br />
    Combining = epic
  </p>
</div>
`;

exports[`should handle multiple paragraphs properly 1`] = `
<div>
  <p>
    React is awesome
And so is markdown
  </p>
  <p>
    Combining = epic
  </p>
</div>
`;

exports[`should handle nested blockquotes 1`] = `
<div>
  <blockquote>
    <blockquote>
      <p>
        Lots of ex-Mootoolers on the React team
      </p>
    </blockquote>
    <p>
      Totally didn't know that.
    </p>
    <blockquote>
      <p>
        There's a reason why it turned out so awesome
      </p>
    </blockquote>
    <p>
      Haha I guess you're right!
    </p>
  </blockquote>
</div>
`;

exports[`should handle ordered lists 1`] = `
<div>
  <ol>
    <li>
      Ordered
    </li>
    <li>
      Lists
    </li>
    <li>
      Are cool
    </li>
  </ol>
</div>
`;

exports[`should handle ordered lists with a start index 1`] = `
<div>
  <ol
    start="7"
  >
    <li>
      Ordered
    </li>
    <li>
      Lists
    </li>
    <li>
      Are cool
    </li>
  </ol>
</div>
`;

exports[`should handle tight, unordered lists 1`] = `
<div>
  <ul>
    <li>
      Unordered
    </li>
    <li>
      Lists
    </li>
    <li>
      Are cool
    </li>
  </ul>
</div>
`;

exports[`should render image references 1`] = `"<p>Checkout out this ninja: <img src=\\"/assets/ninja.png\\" alt=\\"The Waffle Ninja\\"/>. Pretty neat, eh?</p>"`;

exports[`should render link references 1`] = `"<p>Stuff were changed in <a href=\\"https://github.com/rexxars/react-markdown/compare/v1.1.3...v1.1.4\\">1.1.4</a>. Check out the changelog for reference.</p>"`;

exports[`should render tables 1`] = `"<p>Languages are fun, right?</p><table><thead><tr><th style=\\"text-align:left\\">ID</th><th style=\\"text-align:center\\">English</th><th style=\\"text-align:right\\">Norwegian</th><th>Italian</th></tr></thead><tbody><tr><td style=\\"text-align:left\\">1</td><td style=\\"text-align:center\\">one</td><td style=\\"text-align:right\\">en</td><td>uno</td></tr><tr><td style=\\"text-align:left\\">2</td><td style=\\"text-align:center\\">two</td><td style=\\"text-align:right\\">to</td><td>due</td></tr><tr><td style=\\"text-align:left\\">3</td><td style=\\"text-align:center\\">three</td><td style=\\"text-align:right\\">tre</td><td>tre</td></tr></tbody></table>"`;

exports[`should set source position attributes if sourcePos option is enabled 1`] = `
<div
  data-sourcepos="1:1-5:4"
>
  <p
    data-sourcepos="1:1-1:4"
  >
    Foo
  </p>
  <hr
    data-sourcepos="3:1-3:13"
  />
  <p
    data-sourcepos="5:1-5:4"
  >
    Bar
  </p>
</div>
`;

exports[`should skip html blocks if skipHtml prop is set 1`] = `
<div>
  <p>
    This is a regular paragraph.
  </p>
  <p>
    This is another regular paragraph.
  </p>
</div>
`;

exports[`should skip inline html with skipHtml option enabled 1`] = `
<div>
  <p>
    I am having 
    so
     much fun
  </p>
</div>
`;

exports[`should skip nodes that are defined as disallowed 1`] = `
<div>
  <h1>
    Header
  </h1>
  <p>
    Paragraph
  </p>
  <h2>
    New header
  </h2>
  <ol />
  <p>
    Foo
  </p>
</div>
`;

exports[`should skip nodes that are not defined as allowed 1`] = `
<div>
  <p>
    Paragraph
  </p>
  <ol>
    <li>
      List item
    </li>
    <li>
      List item 2
    </li>
  </ol>
</div>
`;

exports[`should unwrap child nodes from disallowed nodes, if unwrapDisallowed option is enabled 1`] = `
<div>
  <p>
    Espen 
    <del>
      initiated
    </del>
     had the initial commit
    , but has had several 
    contributors
  </p>
</div>
`;

exports[`uses passed classname for root component 1`] = `
<div
  className="md"
>
  <p>
    Test
  </p>
</div>
`;
